# Hmwk 1 Part 4 - Homer Kay

In [1]:
#Add packages
%matplotlib inline 
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
#From Scikit Learn
from sklearn import preprocessing
from sklearn.model_selection  import train_test_split, cross_val_score, KFold
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
from sklearn.metrics import roc_curve, auc, confusion_matrix, classification_report

In [2]:
# Checking Current Directory
%pwd

'/Users/homerkay/Desktop/6303 Predictive Models/HMWK 1/Notebooks'

In [3]:
# Change Directory
%cd /Users/homerkay/Desktop/6303 Predictive Models/HMWK 1

/Users/homerkay/Desktop/6303 Predictive Models/HMWK 1


## Reading in Dataset from URL

In [4]:
## Pt 1
pst=pd.read_csv("prostate.csv", sep=",")
## Use Enconding argument, Latin-1, iso... etc to read in files properly
#Had to remove "NA" from dataset native file to get to read properly
pst.head(10)


Obs    lcavol   lweight  age      lbph  svi       lcp  gleason  pgg45  \
0    1 -0.579818  2.769459   50 -1.386294    0 -1.386294        6      0   
1    2 -0.994252  3.319626   58 -1.386294    0 -1.386294        6      0   
2    3 -0.510826  2.691243   74 -1.386294    0 -1.386294        7     20   
3    4 -1.203973  3.282789   58 -1.386294    0 -1.386294        6      0   
4    5  0.751416  3.432373   62 -1.386294    0 -1.386294        6      0   
5    6 -1.049822  3.228826   50 -1.386294    0 -1.386294        6      0   
6    7  0.737164  3.473518   64  0.615186    0 -1.386294        6      0   
7    8  0.693147  3.539509   58  1.536867    0 -1.386294        6      0   
8    9 -0.776529  3.539509   47 -1.386294    0 -1.386294        6      0   
9   10  0.223144  3.244544   63 -1.386294    0 -1.386294        6      0   

       lpsa  
0 -0.430783  
1 -0.162519  
2 -0.162519  
3 -0.162519  
4  0.371564  
5  0.765468  
6  0.765468  
7  0.854415  
8  1.047319  
9  1.047319

## Prostate Target Variable

In [5]:
# Move Prostate column.
# designate target variable name
targetName = 'lpsa'
targetSeries = pst[targetName]
#remove target from current location and insert in collum 0
del pst[targetName]
pst.insert(0, targetName, targetSeries)
#reprint dataframe and see target is in position 0
pst.head(10)

lpsa  Obs    lcavol   lweight  age      lbph  svi       lcp  gleason  \
0 -0.430783    1 -0.579818  2.769459   50 -1.386294    0 -1.386294        6   
1 -0.162519    2 -0.994252  3.319626   58 -1.386294    0 -1.386294        6   
2 -0.162519    3 -0.510826  2.691243   74 -1.386294    0 -1.386294        7   
3 -0.162519    4 -1.203973  3.282789   58 -1.386294    0 -1.386294        6   
4  0.371564    5  0.751416  3.432373   62 -1.386294    0 -1.386294        6   
5  0.765468    6 -1.049822  3.228826   50 -1.386294    0 -1.386294        6   
6  0.765468    7  0.737164  3.473518   64  0.615186    0 -1.386294        6   
7  0.854415    8  0.693147  3.539509   58  1.536867    0 -1.386294        6   
8  1.047319    9 -0.776529  3.539509   47 -1.386294    0 -1.386294        6   
9  1.047319   10  0.223144  3.244544   63 -1.386294    0 -1.386294        6   

   pgg45  
0      0  
1      0  
2     20  
3      0  
4      0  
5      0  
6      0  
7      0  
8      0  
9      0

## EDA

In [6]:
print(pst.shape)
print(pst.dtypes)
print(pst.mean())
print()
print(pst.std())




(97, 10)
lpsa       float64
Obs          int64
lcavol     float64
lweight    float64
age          int64
lbph       float64
svi          int64
lcp        float64
gleason      int64
pgg45        int64
dtype: object
lpsa        2.478387
Obs        49.000000
lcavol      1.350010
lweight     3.628943
age        63.865979
lbph        0.100356
svi         0.216495
lcp        -0.179366
gleason     6.752577
pgg45      24.381443
dtype: float64

lpsa        1.154329
Obs        28.145456
lcavol      1.178625
lweight     0.428411
age         7.445117
lbph        1.450807
svi         0.413995
lcp         1.398250
gleason     0.722134
pgg45      28.204035
dtype: float64


## Test/Train 70/30

In [7]:
features_train, features_test, target_train, target_test = train_test_split(
    pst.iloc[:,1:].values, pst.iloc[:,0].values, test_size=0.30, random_state=0)

In [8]:
print(features_test.shape)
print(features_train.shape)
print(target_test.shape)
print(target_train.shape)

(30, 9)
(67, 9)
(30,)
(67,)


## Ridge

In [9]:
from sklearn.linear_model import Ridge
ridge_reg = Ridge(alpha=0, normalize=True)
ridge_reg.fit(features_train, target_train)
target_predicted_ridge = ridge_reg.predict(features_test)


## Print Coefficients/Metrics

In [10]:
print(ridge_reg.coef_)
mse = np.mean((target_predicted_ridge-target_test)**2)
print('mse')
print(mse)
## Explained Variance Score
from sklearn.metrics import explained_variance_score
print('explained variance score')
print(explained_variance_score(target_test, target_predicted_ridge))  

from sklearn.metrics import mean_absolute_error
print('mean absolute error')
print(mean_absolute_error(target_test, target_predicted_ridge))


[ 0.03699324  0.10517099 -0.11260045 -0.00413638  0.04626408 -0.05959463
  0.02000217 -0.07985978  0.00122541]
mse
0.1339312410276445
explained variance score
0.9141793444812155
mean absolute error
0.2791496668675404


## LASSO Regression

In [27]:
from sklearn.linear_model import Lasso
lassoreg = Lasso(alpha=0.1)
lassoreg.fit(features_train, target_train)
target_predicted_lasso = lassoreg.predict(features_test)


In [28]:
print(lassoreg.coef_)
mse = np.mean((target_predicted_lasso-target_test)**2)
print('mse')
print(mse)
## Explained Variance Score
from sklearn.metrics import explained_variance_score
print('explained variance score')
print(explained_variance_score(target_test, target_predicted_lasso))  

from sklearn.metrics import mean_absolute_error
print('mean absolute error')
print(mean_absolute_error(target_test, target_predicted_lasso))

[ 0.03847472  0.         -0.         -0.          0.          0.
  0.         -0.          0.00053672]
mse
0.1388089804888271
explained variance score
0.911012035962331
mean absolute error
0.2541667102470429


## Elastic Net Regression

In [29]:
from sklearn.linear_model import ElasticNet
en = ElasticNet(alpha=0.1)
en.fit(features_train, target_train)
print(en)
target_predicted_en = en.predict(features_test)


ElasticNet(alpha=0.1, copy_X=True, fit_intercept=True, l1_ratio=0.5,
      max_iter=1000, normalize=False, positive=False, precompute=False,
      random_state=None, selection='cyclic', tol=0.0001, warm_start=False)


In [30]:
print(en.coef_)
mse = np.mean((target_predicted_en-target_test)**2)
print('mse')
print(mse)
## Explained Variance Score
from sklearn.metrics import explained_variance_score
print('explained variance score')
print(explained_variance_score(target_test, target_predicted_en))  

from sklearn.metrics import mean_absolute_error
print('mean absolute error')
print(mean_absolute_error(target_test, target_predicted_en))

[ 0.03823749  0.01140537 -0.         -0.00072161  0.          0.
  0.         -0.          0.00054233]
mse
0.13582239924427414
explained variance score
0.9128079407046956
mean absolute error
0.25312295839349835


## Ordinary Least Squares

In [31]:
from sklearn import linear_model
reg = linear_model.LinearRegression()
ols = reg.fit (features_train, target_train)
target_predicted_reg = reg.predict(features_test)

In [32]:
print(reg.coef_)
mse = np.mean((target_predicted_reg-target_test)**2)
print('mse')
print(mse)
## Explained Variance Score
from sklearn.metrics import explained_variance_score
print('explained variance score')
print(explained_variance_score(target_test, target_predicted_reg))  

from sklearn.metrics import mean_absolute_error
print('mean absolute error')
print(mean_absolute_error(target_test, target_predicted_reg))

[ 0.03699324  0.10517099 -0.11260045 -0.00413638  0.04626408 -0.05959463
  0.02000217 -0.07985978  0.00122541]
mse
0.13393124102764437
explained variance score
0.9141793444812156
mean absolute error
0.2791496668675403


## Part B - Grid Search

In [49]:
# use a full grid over several parameters and cross validate 5 times
from sklearn.model_selection import GridSearchCV
#param_grid = {"alpha": [.01,.1, .5, 1, 2]}
#param_grid={"alpha": [1,10,1]} #this does a range 1 through 10 changes by a factor of 1. 
param_grid={"alpha": [.01,1,.05]} #this does a range 1 through 1 changes by a factor of .05

# run grid search
grid_search = GridSearchCV(ridge_reg, param_grid=param_grid,n_jobs=-1,cv=5)
grid_search.fit(features_train, target_train)
print("Grid Scores", grid_search.cv_results_)
print("Best", grid_search.best_params_)    

Grid Scores {'mean_fit_time': array([0.01380525, 0.00197511, 0.00077386]), 'std_fit_time': array([6.43333332e-03, 9.57701446e-04, 3.19033637e-05]), 'mean_score_time': array([0.00097399, 0.00046082, 0.00035415]), 'std_score_time': array([3.20745371e-04, 1.35377622e-04, 4.38665825e-05]), 'param_alpha': masked_array(data=[0.01, 1, 0.05],
             mask=[False, False, False],
       fill_value='?',
            dtype=object), 'params': [{'alpha': 0.01}, {'alpha': 1}, {'alpha': 0.05}], 'split0_test_score': array([0.88615009, 0.79149991, 0.88713693]), 'split1_test_score': array([0.86381108, 0.57445349, 0.84609634]), 'split2_test_score': array([0.94497562, 0.83255372, 0.95988892]), 'split3_test_score': array([0.79320425, 0.47699615, 0.76610143]), 'split4_test_score': array([0.94899956, 0.83095519, 0.95970145]), 'mean_test_score': array([0.88705655, 0.70074497, 0.88327253]), 'std_test_score': array([0.0567232 , 0.14638414, 0.07230582]), 'rank_test_score': array([1, 3, 2], dtype=int32), 'spli

In [45]:
# use a full grid over several parameters and cross validate 5 times
from sklearn.model_selection import GridSearchCV
param_grid = {"alpha": [.01,.1, .5, 1, 2]}
#param_grid={"alpha": [1,10,1]} #this does a range 1 through 10 changes by a factor of 1. 
#param_grid={"alpha": [.01,1,.05]} #this does a range 1 through 1 changes by a factor of .05

# run grid search
grid_search = GridSearchCV(lassoreg, param_grid=param_grid,n_jobs=-1,cv=5)
grid_search.fit(features_train, target_train)
print("Grid Scores", grid_search.cv_results_)
print("Best", grid_search.best_params_)   

Grid Scores {'mean_fit_time': array([0.00605569, 0.00106015, 0.00079679, 0.00080757, 0.00087118]), 'std_fit_time': array([0.0030463 , 0.00065755, 0.00016081, 0.00027032, 0.00040787]), 'mean_score_time': array([0.00147605, 0.00065145, 0.00064468, 0.00062218, 0.00058947]), 'std_score_time': array([0.000329  , 0.00018447, 0.00037098, 0.00026164, 0.00024844]), 'param_alpha': masked_array(data=[0.01, 0.1, 0.5, 1, 2],
             mask=[False, False, False, False, False],
       fill_value='?',
            dtype=object), 'params': [{'alpha': 0.01}, {'alpha': 0.1}, {'alpha': 0.5}, {'alpha': 1}, {'alpha': 2}], 'split0_test_score': array([0.89728328, 0.90609794, 0.9097718 , 0.91197415, 0.91208025]), 'split1_test_score': array([0.87841618, 0.88007422, 0.88792682, 0.88071235, 0.86330882]), 'split2_test_score': array([0.94247163, 0.93775297, 0.94500014, 0.95356464, 0.96861698]), 'split3_test_score': array([0.8301912 , 0.82733732, 0.82405083, 0.81699191, 0.801157  ]), 'split4_test_score': array([0.

In [46]:
# use a full grid over several parameters and cross validate 5 times
from sklearn.model_selection import GridSearchCV
#param_grid = {"alpha": [.01,.1, .5, 1, 2]}
#param_grid={"alpha": [1,10,1]} #this does a range 1 through 10 changes by a factor of 1. 
param_grid={"alpha": [.01,1,.05]} #this does a range 1 through 1 changes by a factor of .05

# run grid search
grid_search = GridSearchCV(en, param_grid=param_grid,n_jobs=-1,cv=5)
grid_search.fit(features_train, target_train)
print("Grid Scores", grid_search.cv_results_)
print("Best", grid_search.best_params_)   

Grid Scores {'mean_fit_time': array([0.0027638 , 0.00070815, 0.00029507]), 'std_fit_time': array([1.07895461e-03, 4.87605466e-04, 5.14365637e-06]), 'mean_score_time': array([0.00089121, 0.00053406, 0.00020156]), 'std_score_time': array([2.69022169e-04, 4.72494930e-04, 5.33674307e-06]), 'param_alpha': masked_array(data=[0.01, 1, 0.05],
             mask=[False, False, False],
       fill_value='?',
            dtype=object), 'params': [{'alpha': 0.01}, {'alpha': 1}, {'alpha': 0.05}], 'split0_test_score': array([0.89223793, 0.90972197, 0.90333872]), 'split1_test_score': array([0.87468355, 0.88768833, 0.88280911]), 'split2_test_score': array([0.94281548, 0.94537835, 0.93982291]), 'split3_test_score': array([0.81755441, 0.82380176, 0.83211439]), 'split4_test_score': array([0.94664952, 0.93673207, 0.94112402]), 'mean_test_score': array([0.89445004, 0.90060601, 0.8996398 ]), 'std_test_score': array([0.04703741, 0.04288316, 0.03992461]), 'rank_test_score': array([3, 1, 2], dtype=int32), 'spli

## Re-Run Lasso and EN with optimized Alpha

In [34]:
from sklearn.linear_model import Lasso
lassoreg = Lasso(alpha=1)
lassoreg.fit(features_train, target_train)
target_predicted_lasso = lassoreg.predict(features_test)


In [35]:
print(lassoreg.coef_)
mse = np.mean((target_predicted_lasso-target_test)**2)
print('mse')
print(mse)
## Explained Variance Score
from sklearn.metrics import explained_variance_score
print('explained variance score')
print(explained_variance_score(target_test, target_predicted_lasso))  

from sklearn.metrics import mean_absolute_error
print('mean absolute error')
print(mean_absolute_error(target_test, target_predicted_lasso))

[ 0.03751218  0.         -0.         -0.          0.          0.
  0.          0.          0.        ]
mse
0.14100615636069663
explained variance score
0.909603134312027
mean absolute error
0.23862730611581603


In [36]:
from sklearn.linear_model import ElasticNet
en = ElasticNet(alpha=1)
en.fit(features_train, target_train)
print(en)
target_predicted_en = en.predict(features_test)


ElasticNet(alpha=1, copy_X=True, fit_intercept=True, l1_ratio=0.5,
      max_iter=1000, normalize=False, positive=False, precompute=False,
      random_state=None, selection='cyclic', tol=0.0001, warm_start=False)


In [37]:
print(en.coef_)
mse = np.mean((target_predicted_en-target_test)**2)
print('mse')
print(mse)
## Explained Variance Score
from sklearn.metrics import explained_variance_score
print('explained variance score')
print(explained_variance_score(target_test, target_predicted_en))  

from sklearn.metrics import mean_absolute_error
print('mean absolute error')
print(mean_absolute_error(target_test, target_predicted_en))

[ 0.03806647  0.         -0.         -0.          0.          0.
  0.         -0.          0.00019737]
mse
0.1391287322114935
explained variance score
0.9108142390386148
mean absolute error
0.24628887305286906


## Results LASSO/EN
Once again, both Lasso and Elastic Net had almost identical results throughout.  I believe this is because EN is a function of Ridge and Lasso, and Lasso must dominate that function.  Although the difference between Alpha = 0.1 and (optimized thorugh grid search) Alpha = 1 was not as significant on this dataset.  MSE was 14%, Explained Variance 90%, and MAE 24%.  This model is very accurate compared to the Prostate dataset.  

## Part C Cross Validation OLS
Cross Validation suggests the OLS model is not close to optimized and that the variance is not captured well in the model training.  

## Results OLS
Since Alpha = 0 for Ridge Regression is equivalent to OLS there is no difference between the initial Ridge model and this OLS model.  The Cross Validation suggests the model is not stable enough across the dataset to be production ready.  

In [38]:
#verify OLS with Cross Validation
scores = cross_val_score(ols, features_train, target_train, cv=10)
print("Cross Validation Score for each K",scores)
scores.mean()   

Cross Validation Score for each K [0.89429521 0.83005612 0.95197232 0.77815274 0.94589605 0.93673892
 0.72400751 0.79753743 0.92571727 0.94612212]


0.8730495693172788

## Cross Validation RIDGE

In [39]:
#verify Ridge with Cross Validation
scores = cross_val_score(ridge_reg, features_train, target_train, cv=10)
print("Cross Validation Score for each K",scores)
scores.mean()   

Cross Validation Score for each K [0.89429521 0.83005612 0.95197232 0.77815274 0.94589605 0.93673892
 0.72400751 0.79753743 0.92571727 0.94612212]


0.873049569317279

## Optimized RIDGE

In [50]:
from sklearn.linear_model import Ridge
ridge_reg = Ridge(alpha=0.01, normalize=True)
ridge_reg.fit(features_train, target_train)
target_predicted_ridge = ridge_reg.predict(features_test)


In [51]:
print(ridge_reg.coef_)
mse = np.mean((target_predicted_ridge-target_test)**2)
print('mse')
print(mse)
## Explained Variance Score
from sklearn.metrics import explained_variance_score
print('explained variance score')
print(explained_variance_score(target_test, target_predicted_ridge))  

from sklearn.metrics import mean_absolute_error
print('mean absolute error')
print(mean_absolute_error(target_test, target_predicted_ridge))


[ 0.03587394  0.11752058 -0.08737111 -0.00476521  0.04615647 -0.03048378
  0.01331575 -0.07018886  0.00122656]
mse
0.12954080005862953
explained variance score
0.9165333596032441
mean absolute error
0.27105321797768966


In [52]:
#verify Ridge with Cross Validation
scores = cross_val_score(ridge_reg, features_train, target_train, cv=10)
print("Cross Validation Score for each K",scores)
scores.mean()   

Cross Validation Score for each K [0.90163356 0.82815781 0.94835285 0.77540377 0.95336509 0.9428201
 0.71650533 0.7970091  0.93226855 0.95013041]


0.874564656363874

## Results/Conclusions RIDGE (Optimized Model)
Initial Results:
Ridge Alpha = 0 
- MSE - .13
- EVS - .91
- MAE - .28
Grid Search Results 
- Alpha = 0.01
Ridge Alpha = 0.05
- MSE - .13
- EVS - .92
- MAE - .27

During CV I found there were considerable outliers suggesting the full variance of the dataset is not captured during training.  

From Alpha = 0 to Alpha = 0.01 the MSE, EVS, and MAE improved very minimally, suggesting that the penalty method may not be necessary in this solution.  The MAE evaluates the absolute difference in error of the means.  It is easily interpretable.  The explained variance score states how well the variance describes the prediction results.  In this case 92% is very good, but not terrible.  Relating the Variance score to CV you can see that there may be some case of overfitting since the accuracy is getting very good while the CV is showing instability.  